bundle/edu.gemini.spdb.reports.collection/src/main/scala/edu/gemini/dbTools/ephemeris/TcsEphemerisExport.scala

```diff
@@ -2,210 +2,173 @@ package edu.gemini.dbTools.ephemeris
-    }
-  }
+  /** Horizons ids that correspond to ephemeris files that do not need to be
+    * updated because they have sufficient converage for the night already.
+    */
+  final case class UpToDate(hids: ISet[HorizonsDesignation]) extends FileCategory
 
-  type TryExport[A] = EitherT[IO, ExportError, A]
+  /** Categorization of horizons ids according to the disposition of the
+    * corresponding ephemeris file in the directory.
+    */
+  final case class FilePartition(expired: Expired, extra: Extra, missing: Missing, upToDate: UpToDate)
 
-  object TryExport {
-    def fromDisjunction[A](a:ExportError \/ A): TryExport[A] =
-      EitherT(IO(a))
+  def apply(dir: Path, night: Night, site: Site): TcsEphemerisExport =
+    new TcsEphemerisExport {
 
-    def fromTryCatch[A](e: Throwable => ExportError)(a: => A): TryExport[A] =
-      fromDisjunction {
-        \/.fromTryCatchNonFatal(a).leftMap(e)
-      }
-  }
+      import ExportError._
 
-  /** Cron job entry point.  See edu.gemini.spdb.cron.osgi.Activator.
-    */
-  def run(ctx: BundleContext)(tmpDir: File, log: Logger, env: java.util.Map[String, String], user: java.util.Set[Principal]): Unit = {
-    val site = Option(SiteProperty.get(ctx)) | sys.error(s"Property `${SiteProperty.NAME}` not specified.")
+      val files  = EphemerisFiles(dir)
+      val start  = Instant.ofEpochMilli(night.getStartTime)
+      val end    = Instant.ofEpochMilli(night.getEndTime)
 
-    val exportDir = Option(ctx.getProperty(DirectoryProp)).fold(tmpDir) { path =>
-      new File(path)
-    }
+      def partition(hids: ISet[HorizonsDesignation]): TryExport[FilePartition] = {
+        def upToDate(relevant: ISet[HorizonsDesignation]): TryExport[ISet[HorizonsDesignation]] = {
+          // TryExport[ List[(HorizonsDesignation, List[Instant])]  ]
+          val allPairs = relevant.toList.traverseU { hid =>
+            files.parseTimes(hid).strengthL(hid)
+          }
 
-    val odbRef = ctx.getServiceReference(classOf[IDBDatabaseService])
-    val odb    = ctx.getService(odbRef)
-    val night  = TwilightBoundedNight.forTime(NAUTICAL, Instant.now.toEpochMilli, site)
-    val exp    = new TcsEphemerisExport(exportDir, night, site, odb, user)
-
-    log.info(s"Starting ephemeris lookup for $site, writing into $exportDir")
-    try {
-      exp.export.run.unsafePerformIO() match {
-        case -\/(err)     =>
-          ExportError.logError(err, log, "Could not refresh ephemeris data: ")
-
-        case \/-(updates) =>
-          updates.toList.foreach { case (hid, res) =>
-            res match {
-              case -\/(err)  => ExportError.logError(err, log, "")
-              case \/-(file) => log.log(Level.INFO, s"$hid: updated at ${Instant.ofEpochMilli(file.lastModified())}")
+          allPairs.map { lst =>
+            val upToDatePairs = lst.filter { case (hid, times) =>
+              times.filterGt(Some(start)).filterLt(Some(end)).insert(start).insert(end).toList.sliding(2).forall {
```

Comment:
I usually do this by proving the list is non-empty and then zipping it with its own tail. This way you get pairs instead of an arbitrary `List`. No a big deal though.